### PR TITLE
x11: fix vsync support

### PR DESCRIPTION
**Pull request by 27thLiz**
My machine (older intel gpu) doesn't have this extension.
It has `glXSwapIntervalMESA`, tho it crashes when I call that function.
Needs some more investigation..

### CODE REVIEW

File: platform/x11/context_gl_x11.cpp [hunk 1]
```diff
@@ -179,9 +179,18 @@ int ContextGL_X11::get_window_height() {
...
+	static PFNGLXSWAPINTERVALEXTPROC glXSwapIntervalEXT = NULL;
+
+	if (!setup) {
+		setup = true;
+		glXSwapIntervalEXT = (PFNGLXSWAPINTERVALEXTPROC) glXGetProcAddressARB((const GLubyte*)"glXSwapIntervalEXT");
+	}
+	if (glXSwapIntervalEXT) {
+		GLXDrawable drawable = glXGetCurrentDrawable();
+		glXSwapIntervalEXT(x11_display, drawable, p_use?1:0);
+		use_vsync = p_use;
```
Comment (bojidar-bg) [inline platform/x11/context_gl_x11.cpp]: Probably this should get out of the `if`?

File: platform/x11/context_gl_x11.cpp [hunk 2]
```diff
@@ -179,9 +179,18 @@ int ContextGL_X11::get_window_height() {
...
-	//GLXSwapIntervalEXT(x11_display, drawable, p_use?1:0);
-	use_vsync=p_use;
+	static bool setup = false;
+	static PFNGLXSWAPINTERVALEXTPROC glXSwapIntervalEXT = NULL;
+
+	if (!setup) {
+		setup = true;
+		glXSwapIntervalEXT = (PFNGLXSWAPINTERVALEXTPROC) glXGetProcAddressARB((const GLubyte*)"glXSwapIntervalEXT");
+	}
+	if (glXSwapIntervalEXT) {
```
Comment (akien-mga) [inline platform/x11/context_gl_x11.cpp]: Should there be an `else` statement to set `use_vsync=false`?
Also you could probably fix the ``is_using_vsync()`` function right below.

**Comment by J08nY** [comment]
Maybe it would be best to use `glXQueryExtensionsString()` to find out what extension, if any, provides`glXSwapInterval` as per: [opengl.org](https://www.opengl.org/discussion_boards/showthread.php/190765-glXSwapInterval?s=96260eb5d4e1776284580c315f71e0c0&p=1271536&viewfull=1#post1271536)

**Comment by Hinsbart** [comment]
@J08nY seems so, I assumed we'll always get a nullpointer if the method is not supported :/
I still find it strange that it crashes with `glXSwapIntervalMESA` ( `glXQueryExtensionsString()` says it's supported).. mhh

File: platform/x11/context_gl_x11.cpp (hunk 2 above)
Comment (Hinsbart) [inline platform/x11/context_gl_x11.cpp]: IMO it's unnecessary as `use_vsync` can't be true when we don't have the method.

**Comment by zaps166** [comment]
I didn't read everything, but I'm always using this: https://github.com/zaps166/QMPlay2/blob/41b0ff27c3ab8e93e83cec422aa8f97d7333f86c/src/modules/OpenGL2/OpenGL2OldWidget.cpp#L42 (`glXSwapIntervalMESA` as first, `glXSwapIntervalSGI` as a fallback if exists for other drivers) - it works properly on many drivers (also on old Intel). Does `glXSwapIntervalEXT` work properly everywhere? You can also look at SDL2 implementation in `SDL_x11opengl.c`.

File: platform/x11/context_gl_x11.cpp (hunk 2 above)
Comment (akien-mga) [inline platform/x11/context_gl_x11.cpp]: That's right. About the ``is_using_vsync`` function though, it should be fixed to return ``use_vsync`` I guess :)

**Comment by Hinsbart** [comment]
Ok, rebased.
Now it checks the supported extensions and tries the methods in this order: MESA->SGI->EXT.
From what I can tell it should be the same as in the other toolkits, however both the mesa and sgi functions crash here.
So I guess I must have done something wrong in the code.. can anyone tell what's the error?

**Comment by akien-mga** [comment]
It seems to work fine for me on X11 with both Intel / Mesa 12.0.0 RC1 and nouveau (kernel 4.6.1).

**Comment by Hinsbart** [comment]
@akien-mga I found out that the issue was related to the project I was testing with...
If the project has physics 2d thread mode set to multithreaded, it crashes in `glXSwapInterval`...
Multi-threaded rendering however doesn't cause issues... :O
Gonna open an issue about this

**Pull request merged**